libs/execution_context/ebpf_maps.c

```diff
@@ -525,100 +555,122 @@ _delete_per_cpu_hash_map(_In_ ebpf_core_map_t* map)
     ebpf_lock_destroy(&map->lock);
     per_cpu_data = (ebpf_core_per_cpu_data_t*)map->data;
     for (index = 0; index < per_cpu_data->count; index++) {
-        ebpf_hash_table_t** tables = (ebpf_hash_table_t**)&per_cpu_data->data;
-        ebpf_hash_table_destroy(tables[index]);
+        ebpf_core_per_cpu_hash_table_t* tables = (ebpf_core_per_cpu_hash_table_t*)&per_cpu_data->data;
+        ebpf_lock_destroy(&tables[index].lock);
+        ebpf_hash_table_destroy(tables[index].table);
+        tables[index].table = NULL;
     }
     ebpf_free(map);
 }
 
-static ebpf_hash_table_t*
-_get_hash_table_for_cpu(_In_ ebpf_core_map_t* map)
+static ebpf_core_per_cpu_hash_table_t*
+_get_hash_table_for_cpu(_In_ ebpf_core_map_t* map, uint32_t cpu_id)
 {
-    uint32_t current_cpu;
     ebpf_core_per_cpu_data_t* per_cpu_data = NULL;
-    ebpf_hash_table_t** tables;
-    if (ebpf_is_preemptible()) {
-        return NULL;
-    }
+    ebpf_core_per_cpu_hash_table_t* tables;
 
-    current_cpu = ebpf_get_current_cpu();
     per_cpu_data = (ebpf_core_per_cpu_data_t*)map->data;
-    tables = (ebpf_hash_table_t**)&per_cpu_data->data;
-    if (current_cpu < per_cpu_data->count) {
-        return tables[current_cpu];
+    tables = (ebpf_core_per_cpu_hash_table_t*)&per_cpu_data->data;
+    if (cpu_id < per_cpu_data->count) {
+        return tables + cpu_id;
     } else {
         return NULL;
     }
 }
 
 static uint8_t*
-_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, _In_ const uint8_t* key)
+_find_per_cpu_hash_map_entry(_In_ ebpf_core_map_t* map, uint32_t cpu_id, _In_ const uint8_t* key)
 {
     uint8_t* value = NULL;
-    ebpf_hash_table_t* table;
+    ebpf_lock_state_t lock_state;
+    ebpf_core_per_cpu_hash_table_t* table;
     if (!map || !key)
         return NULL;
 
-    table = _get_hash_table_for_cpu(map);
+    table = _get_hash_table_for_cpu(map, cpu_id);
     if (!table)
         return NULL;
 
-    if (ebpf_hash_table_find(table, key, &value) != EBPF_SUCCESS) {
+    lock_state = ebpf_lock_lock(&table->lock);
```

Comment:
Why do you need a lock here?   The value proposition of a per-cpu map is that it's supposed to be lock free.

Reply:
For the user mode API behavior . UM API invoking find - returns an array of values from all cores.

Reply:
We can have concurrent inserts/finds in the hash table. 

Note:
For the hot path case, the cache line backing the lock should be in the CPU associates with this map, so the lock/unlock should be a relatively low cost operation.

Reply:
Note:
Per-cpu maps are not lock-free on Linux. Per-cpu maps on Linux are regular maps with value size = (definition value size + padding) * CPU count. Find, update and delete operations all acquire and release locks.